xds/internal/xdsclient/tlscreds/bundle_test.go

```diff
@@ -0,0 +1,256 @@
+		`{"refresh_interval": "1s"}`,
+		`{"refresh_interval": "1s","ca_certificate_file": "foo"}`,
+		`{"refresh_interval": "1s","certificate_file":"bar","private_key_file":"baz"}`,
+		`{"refresh_interval": "1s","ca_certificate_file":"foo","certificate_file":"bar","private_key_file":"baz"}`,
+	}
+
+	for _, jd := range tests {
+		t.Run(jd, func(t *testing.T) {
+			msg := json.RawMessage(jd)
+			if _, err := NewBundle(msg); err != nil {
+				t.Errorf("NewBundle(%s): expected no error but got: %s", jd, err)
+			}
+		})
+	}
+}
+
+func TestInvalidTlsBuilder(t *testing.T) {
+	tests := []struct {
+		jd, err string
+	}{
+		{`{"ca_certificate_file": 1}`, "failed to unmarshal config: json: cannot unmarshal number into Go struct field .ca_certificate_file of type string"},
+		{`{"certificate_file":"bar"}`, "pemfile: private key file and identity cert file should be both specified or not specified"},
+	}
+
+	for _, test := range tests {
+		t.Run(test.jd, func(t *testing.T) {
+			msg := json.RawMessage(test.jd)
+			if _, err := NewBundle(msg); err.Error() != test.err {
+				t.Errorf("NewBundle(%s): want error %s, got: %s", msg, test.err, err)
+			}
+		})
+	}
+}
+
+type testServer struct {
+	testgrpc.UnimplementedTestServiceServer
+}
+
+func (t testServer) EmptyCall(_ context.Context, _ *testpb.Empty) (*testpb.Empty, error) {
+	return &testpb.Empty{}, nil
+}
+
+func TestCaReloading(t *testing.T) {
+	srvAddr, stopSrv := startServer(t, "localhost:0", tls.NoClientCert)
+
+	serverCa, err := os.ReadFile(testdata.Path("x509/server_ca_cert.pem"))
+	if err != nil {
+		t.Fatalf("Failed to read test CA cert: %s", err)
+	}
+
+	// Write CA certs to a temporary file so that we can modify it later.
+	caPath := t.TempDir() + "/ca.pem"
+	err = os.WriteFile(caPath, serverCa, 0644)
+	if err != nil {
+		t.Fatalf("Failed to write test CA cert: %v", err)
+	}
+	cfg := fmt.Sprintf(`{
+		"ca_certificate_file": "%s",
+		"refresh_interval": ".01s"
+	}`, caPath)
+	tlsBundle, err := NewBundle([]byte(cfg))
+	if err != nil {
+		t.Fatalf("Failed to create TLS bundle: %v", err)
+	}
+
+	conn, err := grpc.Dial(
+		srvAddr.String(),
+		grpc.WithCredentialsBundle(tlsBundle),
+		grpc.WithAuthority("x.test.example.com"),
+	)
+	if err != nil {
+		t.Fatalf("Error dialing: %v", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+	_, err = client.EmptyCall(context.Background(), &testpb.Empty{})
+	if err != nil {
+		t.Errorf("Error calling EmptyCall: %v", err)
+	}
+	// close the server and create a new one to force client to do a new
+	// handshake.
+	stopSrv()
+
+	invalidCa, err := os.ReadFile(testdata.Path("ca.pem"))
+	if err != nil {
+		t.Fatalf("Failed to read test CA cert: %v", err)
+	}
+	// unload root cert
+	err = os.WriteFile(caPath, invalidCa, 0644)
+	if err != nil {
+		t.Fatalf("Failed to write test CA cert: %v", err)
+	}
+
+	// Leave time for the file_watcher provider to reload the CA.
+	time.Sleep(100 * time.Millisecond)
```

Comment:
This could make the test flaky even though the refresh interval is 10ms, Sometimes tests run real slow on GA.

Reply:
OK, I was also concerned about this being flaky, but I don't really know how to address without either making the test very long to run, or building/using something like `https://pkg.go.dev/github.com/stretchr/testify/assert#Eventually` which would run the test until it succeeds, with a timeout. I haven't seen this pattern within the grpc-go codebase so I was hoping you could suggest something.

Reply:
We could create a test context with a reasonable deadline at the top of the test. We do this in a lot of our tests with the deadline being `5s` or `10s`, and we use this context for all APIs that accept a context instead of using `context.Background()`, including RPCs.

```golang
	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
	defer cancel()
```

At the end of the test, instead of sleeping and then making an RPC, we could do the following:
```golang
for ;ctx.Err() == nil; <-time.After(10 * time.Millisecond) {
	// Client handshake should fail because the server cert is signed by an
	// unknown CA.
	_, err = client.EmptyCall(context.Background(), &testpb.Empty{})
	if err == nil {
		// Certs have not yet reloaded. Loop back to try again.
		return
	}
	const wantErr = "certificate signed by unknown authority"
	if status.Code(err) == codes.Unavailable && strings.Contains(err.Error(), wantErr) {
		// Expected error occurs only after successful reloading of certs.
		return
	}
	t.Logf("EmptyCall() failed: %v, wantCode: %s, wantErr: %s", err, codes.Unavailable, wantErr)
}
if ctx.Err() != nil {
	t.Fatalf("Timeout when waiting for certs to be reloaded")
}
```

This also has a sleep hidden inside the `for`. But it does make it a little less timing dependent. When things fail, it will take longer for the test to fail, but in most cases, it will succeed faster than a hard coded sleep.

Reply:
OK, that's essentially what I had in mind with an equivalent of https://pkg.go.dev/github.com/stretchr/testify/assert#Eventually.

I adapted the code to use a similar structure as what you suggest (the server also needs to be recreated between each attempt, otherwise no handshake occur).